src/FeeCollector.sol

```diff
@@ -0,0 +1,228 @@
+
+interface IReceiver {
+    /// Returns current exchange rate and timestamp of publication.
+    /// exchange rate is represented as fixed-point 1e18.
+    function getLatest() external view returns (uint256 _er, uint256 _ts);
+}
+
+interface ICoreContract {
+    function mintFee(uint256 amount) external;
+}
+
+contract FeeCollector is Initializable, UUPSUpgradeable, OwnableUpgradeable {
+    using SafeERC20 for IERC20;
+
+    uint256 private constant ONE = 1e18;
+
+    struct Config {
+        address coreContract;
+        /// Percentage of APY reduction taken as protocol fee.
+        /// Represented in fixed-point 1e18 (1e18 = 1.0).
+        uint256 feeApyReductionPercentage;
+        /// Minimal time between fee collections, in seconds.
+        uint256 collectionPeriodSeconds;
+        /// ERC-20 fee token (maxBTC).
+        IERC20 feeToken;
+        /// Number of decimals of maxBTC (must be <= 18).
+        uint8 maxbtcDecimals;
+    }
+
+    struct State {
+        /// Timestamp of the last successful fee collection.
+        uint256 lastCollectionTimestamp;
+        /// Exchange rate recorded after the last fee collection (fixed-point 1e18).
+        uint256 lastExchangeRate;
+    }
+
+    // Errors
+    error InvalidOwnerAddress();
+    error InvalidRecipientAddress();
+    error InvalidCoreContractAddress();
+    error InvalidFeeTokenAddress();
+    error InvalidFeeReductionPercentage();
+    error InvalidDecimalConversion();
+    error CollectionPeriodNotElapsed();
+    error NegativeOrZeroApy(uint256 currentRate, uint256 lastRate);
+    error InvalidZeroAmount();
+
+    /// @dev keccak256(abi.encode(uint256(keccak256("maxbtc.fee_collector.config")) - 1)) & ~bytes32(uint256(0xff))
+    bytes32 private constant CONFIG_STORAGE_SLOT =
+        0x55a5612efb3791db0b287c4f2521a452b64dc8c1ea03edaa7b8f0870c0bd6300;
+
+    /// @dev keccak256(abi.encode(uint256(keccak256("maxbtc.fee_collector.state")) - 1)) & ~bytes32(uint256(0xff))
+    bytes32 private constant STATE_STORAGE_SLOT =
+        0x966789e57aed537e5e5c5502b1c3700bbababa9893769d9edb5dcfab993bfe00;
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    function initialize(
+        address owner_,
+        address coreContract_,
+        uint256 feeApyReductionPercentage_,
+        uint64 collectionPeriodSeconds_,
+        address feeToken_,
+        uint8 maxbtcDecimals_
+    ) external initializer {
+        if (owner_ == address(0)) revert InvalidOwnerAddress();
+        if (coreContract_ == address(0)) revert InvalidCoreContractAddress();
+        if (feeToken_ == address(0)) revert InvalidFeeTokenAddress();
+        if (
+            feeApyReductionPercentage_ == 0 || feeApyReductionPercentage_ >= ONE
+        ) {
+            revert InvalidFeeReductionPercentage();
+        }
+        if (maxbtcDecimals_ > 18) revert InvalidDecimalConversion();
```

Comment:
Can we not query that from feeToken itself?

Like here: https://github.com/structured-org/erc20-converter/blob/6e83ca998fbdd2b5984a0c539e2802276c919c5e/src/ERC20Converter.sol#L75

Reply:
this parameter is not required anymore